src/ring_theory/dedekind_finite.lean

```diff
@@ -0,0 +1,333 @@
+import linear_algebra.basic
+import data.nat.basic
+import tactic.apply
+import tactic.omega
+import ring_theory.noetherian
+import ring_theory.adjoin
+import data.matrix.basic
+
+namespace dedekind_finite
+
+section
+
+variables (R : Type*)
+
+class dedekind_finite extends ring R :=
+( inv_comm : ∀ a b : R, a * b = 1 → b * a = 1 )
+
+@[priority 100]
+instance dedekind_finite_of_comm_ring [comm_ring R] : dedekind_finite R :=
+⟨λ a b h, h ▸ mul_comm b a⟩
+
+end
+section
+universes u v w
+variable {I : Type u}     -- The indexing type
+variable {f : I → Type v} -- The family of types already equiped with instances
+variables (x y : Π i, f i) (i : I)
+instance pi.dedekind_finite [∀ i, dedekind_finite $ f i] : dedekind_finite (Π i : I, f i) := by pi_instance
+
+end
+section
+universe u
+variables (R : Type u)
+/-  TODO
+instance asubset.ring {S : set R} [is_subring S] : ring S :=
+by apply_instance
+instance asubtype.ring {S : set R} [is_subring S] : ring (subtype S) := subset.ring
+
+instance subring.dedekind_finite [dedekind_finite R] (S : set R) [is_subring S] : dedekind_finite S :=
+by subtype_instance
+ -/
+def is_nilpotent {R : Type*} [ring R] (a : R) := ∃ n : ℕ, a^n = 0
+def nilpotents [ring R] := { a : R | is_nilpotent a }
+
+class reduced extends ring R :=
```

Comment:
I think this should not `extends ring R`. We would get way to many different classes that can't be easily combined without creating even more classes. (We would get `reduced_noetherian` and `reduced_comm_ring` etc.. etc...)
I think it would be better to make `[ring R]` an assumption, and define `reduced` to be a predicate. Possibly it could still be a type class, just like `compact_space X` is a type class that assume `[topological_space X]`.

Reply:
This comment might also apply to dedekind finite rings. I suggest that they don't extend ring, but assume ring as hypothesis.